src/ws/cert-session.c

```diff
@@ -0,0 +1,307 @@
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * This program opens a session for a given user and runs the bridge in
+ * it, using TLS client certificate authentication and mapping it to a user
+ * through sssd (https://www.freeipa.org/page/V4/User_Certificates).
+ */
+
+#include "config.h"
+#include <stdlib.h>
+
+#include <gio/gio.h>
+
+#include "session-utils.h"
+
+static char *last_txt_msg = NULL;
+static char *conversation = NULL;
+
+
+static int
+pam_conv_func (int num_msg,
+               const struct pam_message **msg,
+               struct pam_response **ret_resp,
+               void *appdata_ptr)
+{
+  /* we don't expect (nor can handle) any actual auth conversation here, but
+   * PAM sometimes sends messages like "Creating home directory for USER" */
+  for (int i = 0; i < num_msg; ++i)
+      debug("got PAM conversation message, ignoring: %s", msg[i]->msg);
+  return PAM_CONV_ERR;
+}
+
+static char* sssd_map_certificate (const char *certificate)
+{
+  g_autoptr(GDBusConnection) bus = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GVariant) user_obj_v = NULL;
+  const gchar *user_obj = NULL;
+  g_autoptr(GVariant) username_v = NULL;
+  g_autoptr(GVariant) username_vv = NULL;
+  const gchar *username = NULL;
+
+  /* for output type validation */
+  const GVariantType* vt_object_path[] = { G_VARIANT_TYPE_OBJECT_PATH };
+  g_autoptr(GVariantType) type_object_path_tuple = g_variant_type_new_tuple (vt_object_path, 1);
+
+  bus = g_bus_get_sync (G_BUS_TYPE_SYSTEM, NULL, &error);
+  if (!bus)
+    {
+      g_warning ("Failed to connect to system bus: %s", error->message);
+      exit_init_problem (PAM_AUTHINFO_UNAVAIL);
+    }
+
+  user_obj_v = g_dbus_connection_call_sync (bus,
+                                            "org.freedesktop.sssd.infopipe",
+                                            "/org/freedesktop/sssd/infopipe/Users",
+                                            "org.freedesktop.sssd.infopipe.Users",
+                                            "FindByCertificate",
+                                            g_variant_new ("(s)", certificate),
+                                            type_object_path_tuple,
+                                            G_DBUS_CALL_FLAGS_NONE,
+                                            -1, /* default timeout */
+                                            NULL,
+                                            &error);
+  if (!user_obj_v)
+    {
+      g_warning ("Failed to map certificate to user: %s", error->message);
```

Comment:
This should only be a warning in the case of `!G_DBUS_ERROR_SERVICE_UNKNOWN`. When the lookup succeeds, but doesn't find a matching user and we return `PAM_USER_UNKNOWN`, it should just be a normal info.